src/lib.rs

```diff
@@ -571,3 +582,84 @@ fn do_call_3_args(
     instance.recycle();
     Ok(res?)
 }
+
+/// this is just for init. It is like the 3 args, but take a callback
+/// to return some info on the contract
+fn call_init(
+    cache: *mut cache_t,
+    contract_checksum: Buffer,
+    env: Buffer,
+    info: Buffer,
+    msg: Buffer,
+    db: DB,
+    api: GoApi,
+    querier: GoQuerier,
+    gas_limit: u64,
+    print_debug: bool,
+    contract_info: Option<&mut ContractFlags>,
+    gas_used: Option<&mut u64>,
+    err: Option<&mut Buffer>,
+) -> Buffer {
+    let r = match to_cache(cache) {
+        Some(c) => catch_unwind(AssertUnwindSafe(move || {
+            do_call_init(
+                c,
+                contract_checksum,
+                env,
+                info,
+                msg,
+                db,
+                api,
+                querier,
+                gas_limit,
+                print_debug,
+                contract_info,
+                gas_used,
+            )
+        }))
+        .unwrap_or_else(|_| Err(Error::panic())),
+        None => Err(Error::empty_arg(CACHE_ARG)),
+    };
+    let data = handle_c_error(r, err);
+    Buffer::from_vec(data)
+}
+
+fn do_call_init(
+    cache: &mut Cache<GoApi, GoStorage, GoQuerier>,
+    contract_checksum: Buffer,
+    env: Buffer,
+    info: Buffer,
+    msg: Buffer,
+    db: DB,
+    api: GoApi,
+    querier: GoQuerier,
+    gas_limit: u64,
+    print_debug: bool,
+    contract_info: Option<&mut ContractFlags>,
```

Comment:
Why is this an input and not an output? In contrast to gas_used, we don't need it in the error case.

Reply:
Because we cannot return multiple outputs easily. Otherwise, we need to parse the output (which is raw bytes returned from the contract) and modify it to include the info.

If you have a better idea, we could return the gas info there as well: (gas_info, contract_flags, contract_response)

Reply:
Okay, thanks. No better idea right now.